esp-hal-common/src/analog/adc/cal_line.rs

```diff
@@ -0,0 +1,98 @@
+use core::marker::PhantomData;
+
+use crate::adc::{
+    AdcCalBasic,
+    AdcCalEfuse,
+    AdcCalScheme,
+    AdcCalSource,
+    AdcConfig,
+    Attenuation,
+    RegisterAccess,
+};
+
+/// Marker trait for ADC unit which support line fitting
+///
+/// Usually it means that reference points is stored in efuse.
+/// See also [`AdcCalLine`].
+pub trait AdcHasLineCal {}
+
+// coeff_a is actually a float number
+// it is scaled to put them into uint32_t so that the headers do not have to be
+// changed
+const GAIN_SCALE: u32 = 1 << 16;
+
+/// Line fitting ADC calibration scheme
+///
+/// This scheme implements gain correction based on reference points.
+///
+/// Reference point is a pair of reference voltage and corresponding mean
+/// raw digital ADC value. Such values usually is stored in efuse bit fields
+/// for each supported attenuation.
+///
+/// Also it can be measured in runtime by connecting ADC to reference voltage
+/// internally but this method is not so good because actual reference voltage
+/// may varies in range 1.0..=1.2 V. Currently this method is used as a fallback
+/// (with 1.1 V by default) when calibration data is missing.
+///
+/// This scheme also includes basic calibration ([`AdcCalBasic`]).
+#[derive(Clone, Copy)]
+pub struct AdcCalLine<ADCI> {
+    basic: AdcCalBasic<ADCI>,
+
+    /// Gain of ADC-value
+    gain: u32,
+
+    _phantom: PhantomData<ADCI>,
+}
+
+impl<ADCI: AdcCalEfuse + AdcHasLineCal + RegisterAccess> AdcCalScheme<ADCI> for AdcCalLine<ADCI> {
```

Comment:
Just a code style nit: this would look better as:
```suggestion
impl<ADCI> AdcCalScheme<ADCI> for AdcCalLine<ADCI>
where
    ADCI: AdcCalEfuse + AdcHasLineCal + RegisterAccess
{
```